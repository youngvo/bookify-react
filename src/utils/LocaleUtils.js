import T from 'i18n-react';
import LocationUtils from './LocationUtils';

class LocalUtils {
    static instance = new LocalUtils();

    constructor() {
        if (LocalUtils.instance) {
            return LocalUtils.instance;
        }
        var language = LocationUtils.instance.defaultLanguage();
        var supportedLang = 'de_DE,en_US,es_ES,fr_FR,it_IT,nl_NL,pt_BR';
        if (supportedLang.indexOf(language) < 0) {
            language = 'en_US';
        }
        this._setLocale(language);
        return this;
    }

    setTexts(words) {
      var json = { };
      if (words) {
        this._paserText2Json(json, words);
      }
      T.setTexts(json);
    }

    _setLocale(locale) {
        var json = { };
        var words;
        try {
          let text = require(`!raw!./../assets/locale/${locale}/text.properties`);
          let formatedValue = text.replace(/<b>/g, '*');
          formatedValue = formatedValue.replace(/<\/b>/g, '*');
          formatedValue = formatedValue.replace(/<i>/g, '_');
          formatedValue = formatedValue.replace(/<\/i>/g, '_');
          formatedValue = formatedValue.replace(/<u>/g, '~');
          formatedValue = formatedValue.replace(/<\/u>/g, '~');
          words = formatedValue.split('\n');
          for (var i = 0; i < words.length; i++) {
            words[i] = words[i].replace(/\\n/g, '\r\n');
            words[i] = words[i].replace(/<br[^>]*>/gi, '\r\n');
          }
        } catch (error) {
          throw error;
        }
        if (words) {
          this._paserText2Json(json, words);
        }
        json.urls = { };
        var urls;
        try {
          urls = require(`!raw!./../assets/locale/${locale}/urls.properties`).split('\n');
        } catch (error) {
          throw error;
        }
        if (urls) {
          this._paserText2Json(json.urls, urls);
        }
        if (json) {
          T.setTexts(json);
        }
    }

    _paserText2Json(outputJson, words) {
        var arr = null;
        var key = null;
        var value = null;
        for (var i = 0; i < words.length; i++) {
            arr = words[i].split(':');
            if (arr.length >= 2 && arr[0].indexOf('#') !== 0) {
                key = arr[0].trim();
                arr.shift();
                value = arr.join(':').trim();
                this._decodeNamesToObj(outputJson, key, value);
            }
        }
    }

    _decodeNamesToObj(parentObj, names, value) {
        var arr = names.split('.');
        var name = arr[0];
        if (arr.length === 1) {
            parentObj[name] = { __: value };
            return;
        }
        if (parentObj[name] == null) {
            parentObj[name] = { };
        }
        arr.shift();
        this._decodeNamesToObj(parentObj[name], arr.join('.'), value);
    }

    translate(key, options) {
        if (key == null) {
            return null;
        }
        var result = T.translate(key + '.__' , options);
        if (result === key + '.__') {
          result = key;
        }
        return result;
    }
}

export default LocalUtils;
