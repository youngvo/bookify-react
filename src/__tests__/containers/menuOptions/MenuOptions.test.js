import { shallow, mount, render } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';
import LocaleUtils from './../../../utils/LocaleUtils'
import MenuOptionsContainer, {MenuOptions} from './../../../containers/menuOptions/MenuOptions';
import configureStore from 'redux-mock-store';
import {
    SHOW_CHANGE_BOOK_DESIGN,
    SHOW_CHANGE_ORDER_LAYOUT_SCREEN,
    SHOW_PREVIEW_BOOK,
} from './../../../actions/appStatusActions/RootStatusActions';
import {
    SHOW_MENU_MAIN,
} from './../../../actions/appStatusActions/BookDesignHeaderStatusActions';

const mockStore = configureStore();
let state = {
    photoList: {}
};
describe('MenuOptions Container', () => {
    let wrapper, store;

    beforeEach(() =>{
        store = mockStore(state);
        store.dispatch = jest.fn();
        wrapper = shallow(<MenuOptionsContainer store={store}/>);
    });

    it('maps state and dispatch to props', () => {
        expect(wrapper.props()).toEqual(expect.objectContaining({
            photoList: expect.any(Object),
            onClose:                expect.any(Function),
            onChangeDesignClick:    expect.any(Function),
            onAutoCreateClick:      expect.any(Function),
            onPreviewBookClick:     expect.any(Function),
            onFinishClick:          expect.any(Function),
        }));
    });

    describe('Test mapDispatchToProps()', () => {

        it('maps onClose to dispatch toggleMenuOptions action', () => {
            wrapper.props().onClose();
            expect(store.dispatch).toHaveBeenCalledWith({type: SHOW_MENU_MAIN});
        });

        it('maps onChangeDesignClick to dispatch showChangeBookDesign action', () => {
                wrapper.props().onChangeDesignClick();
                expect(store.dispatch).toHaveBeenCalledWith({type: SHOW_CHANGE_BOOK_DESIGN});
        });

        it('maps onAutoCreateClick to dispatch showChangeOrderLayoutScreen action', () => {
                wrapper.props().onAutoCreateClick();
                expect(store.dispatch).toHaveBeenCalledWith({type: SHOW_CHANGE_ORDER_LAYOUT_SCREEN});
        });

        it('maps onPreviewBookClick to dispatch showPreviewBook action', () => {
                wrapper.props().onPreviewBookClick();
                expect(store.dispatch).toHaveBeenCalledWith({type: SHOW_PREVIEW_BOOK});
        });

        it('maps onFinishClick to dispatch onClose action', () => {
                wrapper.props().onFinishClick();
                expect(store.dispatch).toHaveBeenCalledWith({type: SHOW_MENU_MAIN});
        });
    });

});

describe('test MenuOptions component', () => {
    let menuOptions;
    let photoList = {};
    let onClose;
    beforeEach(() => {
        onClose = jest.fn();
        menuOptions = shallow(<MenuOptions onClose={onClose} photoList={photoList}/>);
    });

    it('MenuOptions Snapshot', () => {
        const comp = renderer.create(<MenuOptions onClose={onClose} photoList={photoList}/>)
        const tree = comp.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('have <div  className="main-menu-component"/>', () => {
        expect(menuOptions.find('div').first().hasClass('main-menu-component')).toBe(true);
        expect(menuOptions.find('.main-menu-component').length).toEqual(1);
    });

    it('have <div  className="triangle-on-menu"/>', () => {
        expect(menuOptions.find('.triangle-on-menu').length).toEqual(1);
    });

    it('have <div  className="main-menu-zone"/>', () => {
        expect(menuOptions.find('.main-menu-zone').length).toEqual(1);
    });

    describe('Test RenderIconButton()', () => {
        it('have className="close-main-menu"/>', () => {
            expect(menuOptions.find('IconButton.close-main-menu').length).toEqual(1);
        });

        it('have dispatch event by Click', () => {
            menuOptions.find('IconButton').simulate('click');
            expect(onClose.mock.calls.length).toEqual(1);
            menuOptions.find('IconButton').simulate('click');
            expect(onClose.mock.calls.length).toEqual(2);
        });
    });

    describe('Test renderMenuOptionsRight()', () => {
        it('have <div className="menu-item-zone menu-item-right">', () => {
            expect(menuOptions.find('.menu-item-zone .menu-item-right').length).toEqual(1);
        });

        it('have <p className="title-text"/>', () => {
            console.log(menuOptions.find('.title-text').last().text());
            expect(menuOptions.find('.menu-item-zone .menu-item-right .title-text').length).toEqual(1);
            expect(menuOptions.find('.menu-item-zone .menu-item-right .title-text').text()).toEqual(LocaleUtils.instance.translate('blurb.com'));
        });

        it('have <div className="sd">', () => {
            expect(menuOptions.find('.menu-item-zone .menu-item-right .sd').length).toEqual(1);
        });

        it('have <LinkText/>', () => {
            expect(menuOptions.find('.menu-item-zone .menu-item-right LinkText').length).toEqual(5);
        });

        it('have dispatch event by LinkText onClick', () => {
            menuOptions.find('.menu-item-zone .menu-item-right LinkText').first().simulate('click');
            expect(onClose.mock.calls.length).toEqual(1);
            menuOptions.find('.menu-item-zone .menu-item-right LinkText').first().simulate('click');
            expect(onClose.mock.calls.length).toEqual(2);
        });

    });

    describe('Test renderMenuOptionsLeft()', () => {
        it('have <div className="menu-item-zone menu-item-left">', () => {
            expect(menuOptions.find('.menu-item-zone .menu-item-left').length).toEqual(1);
        });

        it('have <p className="title-text"/>', () => {
            //console.log(menuOptions.find('.title-text').last().text());
            expect(menuOptions.find('.menu-item-zone .menu-item-left .title-text').length).toEqual(1);
            expect(menuOptions.find('.menu-item-zone .menu-item-left .title-text').text()).toEqual(LocaleUtils.instance.translate('default_book_title'));
        });

        it('have <div className="sd">', () => {
            expect(menuOptions.find('.menu-item-zone .menu-item-left .sd').length).toEqual(2);
        });

        it('have <LinkText/>', () => {
            expect(menuOptions.find('.menu-item-zone .menu-item-left LinkText').length).toEqual(4);
        });

        it('have dispatch event by LinkText onClick', () => {
            let onChangeDesignClick = jest.fn();
            let onAutoCreateClick = jest.fn();
            let onPreviewBookClick = jest.fn();
            let onFinishClick = jest.fn();
            menuOptions.setProps({onChangeDesignClick, onAutoCreateClick, onFinishClick, onPreviewBookClick});
            
            menuOptions.find('.menu-item-zone .menu-item-left LinkText').first().simulate('click');
            expect(onChangeDesignClick.mock.calls.length).toEqual(1);
            
            menuOptions.find('.menu-item-zone .menu-item-left LinkText').at(1).simulate('click');
            expect(onAutoCreateClick.mock.calls.length).toEqual(1);
            
            menuOptions.find('.menu-item-zone .menu-item-left LinkText').at(2).simulate('click');
            expect(onPreviewBookClick.mock.calls.length).toEqual(1);
            
            menuOptions.find('.menu-item-zone .menu-item-left LinkText').last().simulate('click');
            expect(onFinishClick.mock.calls.length).toEqual(1);
        });

    });

});
