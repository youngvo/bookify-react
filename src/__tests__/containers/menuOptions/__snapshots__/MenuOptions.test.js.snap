exports[`test MenuOptions component MenuOptions Snapshot 1`] = `
<div
  className="main-menu-component">
  <div
    className="triangle-on-menu" />
  <div
    className="main-menu-zone">
    <div
      className="menu-item-zone menu-item-left">
      <p
        className="title-text">
        default_book_title
      </p>
      <div
        className="sd" />
      <div
        className="link-text">
        <a
          href={undefined}
          onClick={undefined}
          target="_self">
           
          label.style_book
           
        </a>
      </div>
      <div
        className="link-text">
        <p
          href={undefined}
          style={
            Object {
              "color": "#777777",
            }
          }>
          label.auto_create
        </p>
      </div>
      <div
        className="hr-white">
         
      </div>
      <div
        className="sd" />
      <div
        className="link-text">
        <a
          href={undefined}
          onClick={undefined}
          target="_self">
           
          label.preview_book
           
        </a>
      </div>
      <div
        className="link-text">
        <a
          href={undefined}
          onClick={undefined}
          target="_self">
           
          label.publish
           
        </a>
      </div>
    </div>
    <div
      className="menu-item-zone menu-item-right">
      <p
        className="title-text">
        blurb.com
      </p>
      <div
        className="sd" />
      <div
        className="link-text">
        <a
          href="urls.faq"
          onClick={[Function]}
          target="_blank">
           
          label.faq
           
        </a>
      </div>
      <div
        className="link-text">
        <a
          href="urls.mybooks"
          onClick={[Function]}
          target="_blank">
           
          label.my_blurb_books
           
        </a>
      </div>
      <div
        className="link-text">
        <a
          href="urls.pricing"
          onClick={[Function]}
          target="_blank">
           
          label.pricing
           
        </a>
      </div>
      <div
        className="link-text">
        <a
          href="urls.feedback"
          onClick={[Function]}
          target="_blank">
           
          actions_menu.link.feedback
           
        </a>
      </div>
      <div
        className="link-text">
        <a
          href="urls.customerSupport"
          onClick={[Function]}
          target="_blank">
           
          actions_menu.link.customer_support
           
        </a>
      </div>
    </div>
    <div
      className="close-main-menu icon-button-blurb"
      onClick={[Function]}
      style={
        Object {
          "height": 18,
          "width": 18,
        }
      }>
      <div
        className="icon-CloseIconCleanActive" />
    </div>
  </div>
</div>
`;
